rat/compiler2.py

```diff
@@ -0,0 +1,747 @@
+from dataclasses import dataclass, field
+import functools
+import numpy
+import pandas
+import jax
+import jax.numpy
+import jax.scipy.stats
+from typing import List, Dict, Set, Union
+import itertools
+
+import pandas as pd
+
+from . import ast
+from .codegen_backends import IndentWriter, BaseCodeGenerator, DiscoverVariablesCodeGenerator
+from .exceptions import CompileError, MergeError
+from .variable_table import VariableTracer, VariableTable, VariableType
+from .subscript_table import SubscriptTable
+from .walker import RatWalker, NodeWalker
+from rat import variable_table
+from .position_and_range import Position, Range
+from rat import subscript_table
+
+
+@dataclass
+class StatementInfo:
+    statement: ast.Statement
+    primary: ast.Variable
+
+
+def _get_subscript_names(node: ast.Variable) -> List[str]:
```

Comment:
This is confusing since we're not running this on statements but on variable nodes.

Reply:
Triple nested walkers lol

Reply:
Looool, yeah I went kinda nuts with the little walker things. I quite liked them, they seem very modular in a useful way. 

Like, even if sometimes stuff feels a little old-fashion (passing arguments through member variables), it also feels very self-contained and understandable.

Before (and still now) we have this giant compiler class that has a bunch of functions that all sorta come together and do work. With the walkers it seems like we could just put each walker in its own file in a folder called "Walkers" and then call them from other places and then the compiler code itself is smaller, which seems nice. Anyway, that could have its own drawbacks lol.